File: core/session.py
"""Extract car, track, driver, and the full CarSetup block from .ibt session YAML."""
import re
import yaml


def _safe_yaml(text):
    # iRacing session YAML can contain unquoted values with ':' or non-ASCII;
    # sanitize the known offenders before parsing.
    cleaned = []
    for line in text.splitlines():
        if re.match(r'\s*(DriverSetupName|UserName|TeamName|AbbrevName|Initials|TrackName|TrackDisplayName|TrackDisplayShortName|TrackCity|TrackCountry|SessionName|SessionType|SessionSubType|WeekendOptions)\s*:', line):
            key, _, val = line.partition(':')
            val = val.strip().replace('"', "'")
            line = f'{key}: "{val}"'
        cleaned.append(line)
    return yaml.safe_load('\n'.join(cleaned))


def parse_session_info(session_info: str) -> dict:
    """Returns {car, car_id, track, track_id, driver, session_type, setup: {flat param dict}, setup_name}."""
    try:
        doc = _safe_yaml(session_info)
    except yaml.YAMLError:
        doc = None
    if not isinstance(doc, dict):
        return {'car': 'unknown', 'track': 'unknown', 'driver': '', 'setup': {}, 'setup_name': '', 'session_type': ''}

    weekend = doc.get('WeekendInfo') or {}
    track = str(weekend.get('TrackDisplayName') or weekend.get('TrackName') or 'unknown')
    track_id = weekend.get('TrackID', 0)

    car, car_id, driver = 'unknown', 0, ''
    di = doc.get('DriverInfo') or {}
    idx = di.get('DriverCarIdx', 0)
    for d in di.get('Drivers') or []:
        if d.get('CarIdx') == idx:
            car = str(d.get('CarScreenName') or d.get('CarPath') or 'unknown')
            car_id = d.get('CarID', 0)
            driver = str(d.get('UserName') or '')
            break

    setup_name = str(di.get('DriverSetupName') or '')
    setup = flatten_setup(doc.get('CarSetup') or {})

    session_type = ''
    sessions = (doc.get('SessionInfo') or {}).get('Sessions') or []
    if sessions:
        session_type = str(sessions[-1].get('SessionType') or '')

    return {'car': car, 'car_id': car_id, 'track': track, 'track_id': track_id,
            'driver': driver, 'setup': setup, 'setup_name': setup_name,
            'session_type': session_type}


def flatten_setup(car_setup: dict) -> dict:
    """CarSetup YAML tree -> flat {'Tab.Section.Param': 'value string'} dict."""
    flat = {}

    def walk(node, path):
        if isinstance(node, dict):
            for k, v in node.items():
                if k == 'UpdateCount':
                    continue
                walk(v, path + [str(k)])
        else:
            flat['.'.join(path)] = str(node)

    walk(car_setup, [])
    return flat


_NUM_RE = re.compile(r'-?\d+(?:\.\d+)?')


def numeric_value(raw: str):
    """First numeric token of a setup value string ('55.0%', '-2.8 deg', '6 clicks')."""
    m = _NUM_RE.search(str(raw))
    return float(m.group()) if m else None

File: docs/js/session.js
// Session-YAML metadata + CarSetup extraction — port of core/session.py.

import * as yamlMod from './vendor/js-yaml.min.js';

// UMD resolves differently by host: browsers get globalThis.jsyaml from the
// side-effect; Node's CJS detection routes it through module.exports instead.
const yaml = globalThis.jsyaml || yamlMod.default || yamlMod;

const QUOTE_KEYS = /^(\s*)(DriverSetupName|UserName|TeamName|AbbrevName|Initials|TrackName|TrackDisplayName|TrackDisplayShortName|TrackCity|TrackCountry|SessionName|SessionType|SessionSubType)(\s*):(.+)$/;

function safeYaml(text) {
  try {
    const doc = yaml.load(text);
    if (doc && typeof doc === 'object') return doc;
  } catch { /* fall through to sanitizer */ }
  const cleaned = text.split('\n').map((line) => {
    const m = line.match(QUOTE_KEYS);
    if (m && m[4].trim()) {
      return `${m[1]}${m[2]}: "${m[4].trim().replace(/"/g, "'")}"`;
    }
    return line;
  }).join('\n');
  return yaml.load(cleaned);
}

export function flattenSetup(carSetup) {
  const flat = {};
  (function walk(node, path) {
    if (node && typeof node === 'object' && !Array.isArray(node)) {
      for (const [k, v] of Object.entries(node)) {
        if (k === 'UpdateCount') continue;
        walk(v, path ? `${path}.${k}` : k);
      }
    } else {
      flat[path] = String(node);
    }
  })(carSetup || {}, '');
  return flat;
}

export function numericValue(raw) {
  const m = String(raw).match(/-?\d+(?:\.\d+)?/);
  return m ? parseFloat(m[0]) : null;
}

export function parseSessionInfo(sessionInfo) {
  let doc = null;
  try { doc = safeYaml(sessionInfo); } catch { /* ignore */ }
  if (!doc || typeof doc !== 'object') {
    return { car: 'unknown', track: 'unknown', driver: '', setup: {}, setup_name: '', session_type: '' };
  }
  const weekend = doc.WeekendInfo || {};
  const track = String(weekend.TrackDisplayName || weekend.TrackName || 'unknown');

  let car = 'unknown', driver = '';
  const di = doc.DriverInfo || {};
  const idx = di.DriverCarIdx ?? 0;
  for (const d of di.Drivers || []) {
    if (d.CarIdx === idx) {
      car = String(d.CarScreenName || d.CarPath || 'unknown');
      driver = String(d.UserName || '');
      break;
    }
  }
  let sessionType = '';
  const sessions = (doc.SessionInfo || {}).Sessions || [];
  if (sessions.length) sessionType = String(sessions[sessions.length - 1].SessionType || '');

  return {
    car, track, driver,
    setup: flattenSetup(doc.CarSetup || {}),
    setup_name: String(di.DriverSetupName || ''),
    session_type: sessionType,
  };
}
